### ISSUE TITLE
Don't lock in a specific version of Lucene in the help feature

### ISSUE THREAD

**Issue by merks** [eclipse-platform/eclipse.platform.releng#230]
The help feature includes Lucene bundles.

https://github.com/eclipse-platform/eclipse.platform.releng/blob/acb96341f36482d04898c9927bfa11835d337f30/features/org.eclipse.help-feature/feature.xml#L59-L79

This locks in a very specific version.  The only bundle that actually requires Lucene allows a range of versions:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/31a9e894-2227-4f42-aba4-405b2af03d3d)

Does anything speak against removing the includes?

**Comment by merks** [comment]
The problem is that we've been on the "everyone use the latest" path and now people want to to use 9.5.0, e.g., Mylyn wants that.    Of course people also want the Platform in their target platform and they want to be able to launch a self-hosted IDE, which means they have multiple versions of Lucene in their target platform, which is a complete nightmare mess...

Anyway, let's first see if this breaks anything...

**Comment by laeubi** [comment]
I might be wrong but last time I suggested such changes it was argued that:

1. source bundles might not be included
2. bundle might not be included in update-sites

but if this is no longer an issue I think actually *any* third party dependency can be removed from platform features making them much more flexible (less version bumps and so on...)

**Comment by merks** [comment]
Includes have been removed all over the place.  That's been the general trend.  It's even been suggested that all 3rd party dependencies be removed from includes in general.   In other projects I've created a "dependencies"feature and put 3rd party bundles there to include them easily in the update site without including them in any user-installed feature...

**Comment by merks** [comment]
FYI, in this issue we are discussing the details around using an m2e target location to create wrappers for Lucene 9.5.0...

https://github.com/merks/simrel-maven/issues/5

Will the Platform be able/willing to upgrade to such a new version of Lucene?

**Comment by akurtakov** [comment]
Lucene upstream jars are not osgi so still Orbit wrappers are used and I've been the one doing the last few. So if there are already created wrappes that would  be more than welcome.

**Comment by laeubi** [comment]
Is orbit on github already? If not won't it be better to move there, enable dependabot, getting dependency updates notification, and so on? Or is it now the goal to replace orbit with some "oomph maven wrapped bundles" service?

**Comment by merks** [comment]
I'm waiting for these reviews to complete, though the Platform doesn't use these specifically:

https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8488
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8486

But yes, I already have drop-in replacement wrappers for Lucene 9.5.0 that I've been testing with Mylyn and with DataTools.

I'm testing with a self hosted launch with these new Lucene wrappers, but there are other problems that seem unrelated:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/932405f0-f872-4b69-a46f-8d9474e31ef7)

The Help view appears to work, but I don't know if that tells me anything...

**Comment by merks** [comment]
Orbit is here:

  https://github.com/eclipse/orbit

But they want to use this organization:

https://github.com/eclipse-orbit

Let's discuss goals for Orbit elsewhere.  Here are their issues:

**Comment by merks** [comment]
It sounds like you want to discuss https://github.com/eclipse/orbit/issues/8 here where few will see it so I suggest you discuss Orbit's plans on that Orbit issue where many where many will see the discussion.

**Comment by HannesWell** [comment]
I also think we should remove the lucense bundles from the feature.

> Lucene upstream jars are not osgi so still Orbit wrappers are used and I've been the one doing the last few. So if there are already created wrappes that would be more than welcome.

The issue to request that is now on GH: https://github.com/apache/lucene/issues/8573

It looks like lucene removed the split-packages with version 9:
https://github.com/apache/lucene/blob/main/lucene/MIGRATE.md#migration-from-lucene-8x-to-lucene-90
This would make upstream metadata much simpler.
From my preliminary analysis the only difficulty left is the use of ServiceLoader, but this can be tackeled with the usual ServierLoader Medaitor approach. Or one could use fragments, but I have not yet checked if that fits well into the existing bundle structure.
The biggest problem I have, is that they use Gradle and I don't know Gradle. Does anybody of you know it and can help to configure the bnd-gradle-plugin?

**Comment by merks** [comment]
This locking in of a specific gson version is also annoying:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/db67b80a-1ddb-4c5c-aaa3-26078150f1a8)

This is made worse by the fact that the metadata of the direct-from-maven bundle is different from that of the orbit wrapped bundle:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/1148aef8-c1c8-4f2b-8085-12afba246295)

And lsp4j appears to be relying on the internal packages.  But that's another problem...

**Comment by laeubi** [comment]
> This locking in of a specific gson version is also annoying:

Also I'm quite sure you are aware of it I just wanted to note for others that no one is really "locking" a version in a feature but this is an intrinsic feature of how `feature.xml` works, what is that _included_ plugins/features are resolved (and replaced) by the version of the resolved build time artifact if not already specify a hard version requirement.

Last time I'd attempted to change that it resulted in a lot of complaints and was claimed that no one needs this and it behaves like every one wants it, so its a bit strange to complain now about this, the only solution would be to strip the gson feature completely...

**Comment by merks** [comment]
Yes indeed, features are doing exactly what they are intended to do.   I don't think we should change that because that's how an include of a bundle in a features is expected to work and how it's supposed to work. 

In this issue I'm not complaining about the behavior---I'm not even complaining---I'm suggesting to remove the include, and asking what speaks against that.

It's been suggested in general to remove includes of 3rd party bundle includes from features:

https://github.com/eclipse/orbit/issues/8#issuecomment-1534255305

As you've noted previously, now that Tycho supports many more useful things such as automatically including source bundles for otherwise included non-source bundles there seems to be less that speaks against this approach.  (I suppose there might remain a concern about whether specific sources are installed in the Eclipse SDK, but no one has mentioned that.)  Often other projects want specific 3rd party dependencies in their update site, but that can be accomplished by specifying the bundles directly in the category.xml or by creating a dependencies feature (one that the user does not install), including the bundles there, and including that in the category.xml.  I have taken these approaches with several other projects, and specifically recently with Mylyn...

But I fully understand @akurtakov concern that it's rather late in the cycle to alter this without risk.

**Comment by mickaelistria** [comment]
> I'm suggesting to remove the include, and asking what speaks against that.

I support this approach. Inclusion in features now seems like a pre-p2 utility when features where distributed as zips to unpack in the eclipse installation; but with p2 fetching transitive deps and Platform building its repo with `includeAllDependencies` to include all these transitive deps, feature inclusions are for Platform most often a source of problems and extra-work than a help.
Removing the problem is 10x more valuable than trying to workaround it.

**Comment by laeubi** [comment]
> I'm suggesting to remove the include, and asking what speaks against that.

Nothing special, but if you remove the gson bundle from the gson feature whats left there? :-)

**Comment by merks** [comment]
@mickaelistria 

Indeed, I've grown tried of doing qualifier bundles when updating the *.target!

@laeubi 

There is no gson feature there is a tips feature:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/99ba359a-0822-4ad9-ac96-dacee5426a4e)

@akurtakov 

Yes, I wondered about the indexing, because I've seen issues about such warnings.  Consider this though:

![image](https://github.com/eclipse-platform/eclipse.platform.releng/assets/208716/39a03ac4-4d59-4d00-9e83-c5aa3957a3c9)

Just because the feature requires to install a specific version does not mean that the bundle itself will be wired to that version.  It's happy with either version.  So yes, I think it would be better to stop pre-indexing even if we didn't remove the include...

**Comment by akurtakov** [comment]
Yes, you're correct about the resolution of lucene to newer version if available, still I find it a bit risky to do now as the Help system is not fun to debug and fix at all if smth breaks.
My vote goes for stop pre-indexing, deprecate the extension point and remove Lucene from the feature.xmls for 2023-09.

**Comment by laeubi** [comment]
Alright I was confused because it uses an icon very similar to the feature icon here:
![grafik](https://github.com/eclipse-platform/eclipse.platform.releng/assets/1331477/d8cf740c-114a-4fee-9eec-68e446faa530)

**Comment by merks** [comment]
@laeubi 

No problem!  There's so much detail!!

@akurtakov 

I agree.  We've lived with Lucene duplicates for a long time, so it's safest to wait for the next release rather to change things during RC1.

**Comment by merks** [comment]
FYI,

I will meet with @jonahgraham next week to discuss Orbit.   Then we can hopefully build new versions of Lucene 9.6 which we can contribute to the target platform...

**Comment by HannesWell** [comment]
> I will meet with @jonahgraham next week to discuss Orbit. Then we can hopefully build new versions of Lucene 9.6 which we can contribute to the target platform...

Sounds good, I assume this also includes your new Maven-SimRel-Orbit and the general handling of Maven-targets?
Looking forward to the results. I already thought about, if it would be helpful to provide a little presentation/demo on how to set up/use Maven target properly in an Eclipse project that contributes to the SimRel.
Maybe this would convince more projects to use Maven-Targets directly.
